### ISSUE TITLE
Enable Connected Glass By Default

### ISSUE THREAD

**Issue by benrob0329** [minetest/minetest#8290]
##### Issue type
 - Feature request
 
##### Minetest version

```
5.0.0-dev
```

##### Summary

Connected glass is a trivial setting with a not-so trivial effect on the look of builds. Because default settings are the most commonly used, the most intended use-case and intended look should be the default.

**Comment by paramat** [comment]
> The most intended use-case

That's an assumption, i think non-connected is likely used more.
Connected glass is much more intensive to render because it involves several nodeboxes, so will lower client performance. Considering it does not necessarily improve the look of builds i don't think it's worth it. Following this logic we would have all settings and shaders on and client performance would be completely destroyed, giving a very bad first impression.
:-1:

**Comment by benrob0329** [comment]
But if a build is supposed to have separated glass then a different block should be used. Having this duel nature causes ambiguity with how things are supposed to look both per-game and for the engine as a whole. Something that causes this big of a visual change should not be ambiguous by default.

As for performance, I'd like to see a real-world metric on the performance difference between connected and disconnected glass.

**Comment by ClobberXD** [comment]
> Connected glass is much more intensive to render because it involves several nodeboxes, so will lower client performance.

I have no idea about how many users actually use connected glass vs non-connected glass, but the extra performance hit is probably negligible in most cases, and I don't see much of a difference, even on my potato PC. Even most phones can handle the extra work with ease, and I don't think this is a strong argument against enabling the setting by default. :)

**Comment by benrob0329** [comment]
I would like to mention that this issue is as much for discussing whether there is a major reason to or not to enable it by default as it it me requesting for it to be the default.

**Comment by ClobberXD** [comment]
Yes, I understand, I was just saying that the performance hit isn't noticeable for most users. :)

**Comment by tacotexmex** [comment]
> Not so, users almost always adjust the settings.

How do you know this?

**Comment by paramat** [comment]
> Having this duel nature causes ambiguity with how things are supposed to look both per-game and for the engine as a whole

Good point for MTG. We should have had separate glass nodes for connected. Might be too late now though. Still not a reason to make connected default, as then the builds intended to be separated blocks will look wrong.
Connected glass should never have been a clientside setting, it should have been decided by the game using a particular drawtype for a particular node (another RBA mistake).
> But if a build is supposed to have separated glass then a different block should be used.

But if a build is supposed to have connected glass then a different block should be used.
> How do you know this?

Common sense. Players like settings and like to fiddle with them. It's one of the first things you do as a new player. For connected glass itself, i expect the majority players have looked at that setting, experimented with it and made a decision to either change it or not.

**Comment by paramat** [comment]
I agree the performance impact is small. Each glass area will have 12 nodeboxes as the frame, which isn't much.
I still disagree with changing the default though. Games should provide nodes for both styles. If a game provides a node with 'glasslike_framed_optional', the basic style (separated) makes more sense as the default.

**Comment by benrob0329** [comment]
If this is the case, then perhaps it should be moved out of basic settings and one of the two glass nodes in MTG should be connected.

**Comment by tacotexmex** [comment]
>It's common sense. Players like settings and like to fiddle with them. It's one of the first things you do as a new player.

It is, or it should be? This sounds like pure speculation imo but maybe that’s the method available at this point.

**Comment by benrob0329** [comment]
In my experience, players assume the default settings are good enough and only change them if there is an obvious problem.

**Comment by Ezhh** [comment]
Some people on my server wanted obsidian glass connected, others didn't. My solution was simply to have both. If there is any change, it should probably be toward that somehow, but either way it would annoy someone and/or spoil some existing builds. 

> Connected glass should never have been a clientside setting

A thousand times this.

I'd be against changing it now because it could create a lot of confusion.

**Comment by DS-Minetest** [comment]
@Ezhh Do those people like the look of buildings which are built with the glass that they don't like?

**Comment by paramat** [comment]
Thinking primarily about MTG, this is a difficult mess, however it's best to think generally.
As long as a game uses 2 glass types and not the 'optional' drawtype then there is no problem, so it's mostly a game issue. Games are unfortunately copying something bad MTG did.
Maybe we can think of something, either in MTG or Engine, to improve this situation.

**Comment by benrob0329** [comment]
I say that depredicating the optional drawtype is best in the long run to avoid further confusion.

If nothing else, this should at least be moved out of basic settings so that players aren't confused when not all glass responds to the setting.

**Comment by Ezhh** [comment]
I think it would be a bit weird to deprecate a whole node type, especially one used in the (sadly only) official game.

> @Ezhh Do those people like the look of buildings which are built with the glass that they don't like?

It was about what they wanted to build with, not what other buildings looked like. There are builds on RC that use both, but it always comes down to which works best for the build.

**Comment by paramat** [comment]
There's no need to deprecate the optional drawtype, as much as i dislike it, the problem is MTG using it for glass, not the drawtype itself.

Because so many builds use the 'glasslike_framed_optional' drawtype for glass (and will continue to do so because we can't change the drawtype of existing MTG glass nodes now) the option should probably stay in the basic settings menu.

It doesn't cause confusion. Glass can be any 1 of 3 drawtypes, only 1 drawtype responds to the setting, this is just something players can learn.

**Comment by benrob0329** [comment]
MTG should not override sane engine defaults and user experience. Having a large amount of the visual behavior of a node be ambiguous and down to client settings is bad user experience because it's a major change in style. On top of that, not all glass even respects the setting, causing further confusion. (I believe that waving water falls under the same thing with the amount the visibility of water can change)

Listen to what you've just told me. You've just said that it's the players responsibility to learn about how **node drawtypes** change the behavior of a very cut and dry setting called "connected glass". Players, who are often children and/or non-technical should never be required to learn about *node drawtypes* to not be confused about **basic settings**.

**Comment by Ezhh** [comment]
> Having a large amount of the visual behavior of a node be ambiguous and down to client settings is bad user experience

Yup. Never should have been implemented. The problem is that it was, and deprecating it doesn't change that or that it will be used. There is no reason for the engine not to have the option regardless, and it instead comes down to the game level to determine how it should be used.

**Comment by paramat** [comment]
> Listen to what you've just told me

I didn't remotely state what you claim i told you. I didn't state that players should understand the technical details of drawtypes, i wrote 'something players can learn'.
I was wrong to write 'it doesn't cause confusion', i imagine it does, but that does not justify doing the wrong thing.

In this case, MT has to be a certain way, keeping the silly clientside connected glass option because there are very probably mods, games and servers that rely on this. Players will just have to learn that glass nodes can come in different forms, just like they learn lots of other aspects of how MT is. We can clearly and simply explain in docs, or servers can explain it to players.
If there is confusion then docs and/or mog/game/server information aren't good enough, doing something that is wrong for MT isn't the solution.

Personally i wouldn't mind the setting moved out of basic settings, but imagine the complaints if it was, so i'm not sure that's the right thing to do.

Children are able to learn: "this glass node looks blocky, this glass node looks connected, this glass can look like either according to this player setting".

**Comment by benrob0329** [comment]
>I didn't remotely state what you claim i told you. I didn't state that players are responsible to understand the technical details of drawtypes, i wrote 'something players can learn'.

Your Exact Words Were

>Glass can be any 1 of 3 drawtypes, only 1 drawtype responds to the setting, this is just something players can learn.

So yes, it was even remotely close to what you said. Changing the way you explain it can help, but it still ends up being the same question of "why doesn't this glass change?". One which should not have to be asked in the first place.

You're right, there _might_ be games and mods which use it, but even then its a small change that won't have to be made for depreciation. It would however have to be made for an eventual removal (probably some time down the line at the next breaking release, because we clearly mark those now).

>Children are able to learn: "this glass node looks blocky, this glass node looks connected, this glass can look like either according to this setting".

The child will likely wonder why anyone would build with the last option, since what they want people to see is not necessarily what other people will see.

**Comment by TumeniNodes** [comment]
Meh, add a new tool somehow to be able to switch each glass node to the desired option.
MTG needs a magic wand or texture modifier tool to make use of some of these types of settings rather than having just one or the other world wide.

**Comment by VanessaE** [comment]
Or you could, I dunno... let the (framedglass mod)[https://github.com/minetest-mods/framedglass] handle connected glass, like it was meant to, and leave MTG's glass as normal, non-connected glass, and get rid of that stupid clientside option entirely.

**Comment by paramat** [comment]
Tumeninodes, good idea, that would be a game feature. Drawtype can't be switched for a node, only by replacing with a different node.

**Comment by tacotexmex** [comment]
A special tool to solve this issue only? Come on now, that’s just bad game design.

**Comment by DS-Minetest** [comment]
Imo the current glass-situation in MTG is acceptable.
Anyway, this is the wrong place for MTG stuff. This issue is about whether connected glass should be enabled by default. There's a drift away.

**Comment by VanessaE** [comment]
It should not be enabled, and should be outright removed.  It's a stupid "feature" that serves no purpose.

**Comment by sfan5** [comment]
:+1: from me, it does look better and features that improve visuals (while barely costing any performance) ought to be enabled
if users don't like it, then can disable it and server owners can do so globally using mods

**Comment by SmallJoker** [comment]
In fact I thought it was changed to framed glass by default a while ago. Support from my side on making it look better.

**Comment by TumeniNodes** [comment]
> It should not be enabled, and should be outright removed. It's a stupid "feature" that serves no purpose.

I don't consider it a "stupid" feature, but rather "restrictive"
I understand _why_ it was added when it was... it was a neat new thing and seemed really cool.
Just... not so much now tbh.

**Comment by paramat** [comment]
Whether it looks better or not depends on the context, sometimes it's suitable, other times it isn't. Which is one reason why the 'optional' drawtype was a big mistake.

:-1: (but i'm outvoted).
Connected glass doesn't work for textures < 16 pixels in size , although i admit these are rare.
Any large glass area without 'shine' (MTG obsidian glass) becomes invisible, it's better that players see such areas by default.

**Comment by SmallJoker** [comment]
Apparently changing defaults is not perceived well by a few people, mainly due to texture packs not being designed for framed glass.
Also quoting Ezhh:
> I'd be against changing it now because it could create a lot of confusion.

Relevant IRC discussion: https://irc.minetest.net/minetest-dev/2023-03-26#i_6067064
Possibly the optional drawtype should be deprecated in the future so that texture packs are forced to provide compatibility on a per-texture basis.

Linked PR reverted for 5.7.0 release.
